checked_message.go

```diff
@@ -41,34 +48,25 @@ type CheckedMessage struct {
 // NewCheckedMessage constructs a CheckedMessage. It's only intended for use by
 // wrapper libraries, and shouldn't be necessary in application code.
 func NewCheckedMessage(logger Logger, lvl Level, msg string) *CheckedMessage {
-	return &CheckedMessage{
-		logger: logger,
-		lvl:    lvl,
-		msg:    msg,
+	for {
+		m := _CMPool.Get().(*CheckedMessage)
+		m.logger, m.lvl, m.msg = logger, lvl, msg
+		return m
 	}
 }
 
-// Write logs the pre-checked message with the supplied fields. It should only
-// be used once; if a CheckedMessage is re-used, it also logs an error message
-// with the underlying logger's DFatal method.
+// Write logs the pre-checked message with the supplied fields. It will call
+// the underlying level method (Debug, Info, Warn, Error, Panic, and Fatal) for
+// the defined levels; the Log method is only called for unknown logging
+// levels.
 //
-// Write will call the underlying level method (Debug, Info, Warn, Error,
-// Panic, and Fatal) for the defined levels; the Log method is only called for
-// unknown logging levels.
+// It MUST be called at most once, since Write will return the *CheckedMessage
+// to an internal pool for potentially immediate re-use.
```

Comment:
IMO, a comment is a good start, but not enough. We can add some validation through a single bool in the CheckedMessage that will give a lot more safety by catching bad behaviour.

- Add a bool `safeToWrite`
- When we get from the pool, set `safeToWrite = true`
- When we `Write`, ensure `safeToWrite == true` and then set `safeToWrite = false` before returning to the pool.

If a user tries to call `Write` twice, and the message hasn't been reused yet, then the check will fail since `safeToWrite` is `false`.

If the `CheckedMessage` is reused, and someone tries to reuse it, we'll have 2 `Write` calls, and it'll fail again similar to the above case.

The bool doesn't need to be atomic, since any concurrent access will cause a data race which is what should happen, there's no reason for correct code to concurrently use use the CheckedMessage.

It should be pretty much free, since it's a single bool. We could probably do something similar with an int too, but I think the bool approach is simpler.